### PR TITLE
executor/dirtools: ensure UploadTree set result symlinks

### PR DESCRIPTION
**Pull request by sluongng**
Currently we only support OutputDirs and OutputFiles. Therefore actions
which produce only symlinks would fail when executed remotely with
BuildBuddy.

See https://github.com/bazelbuild/rules_go/pull/3539 as an example.

Implement support for OutputFileSymlinks and OutputDirectorySymlinks.

**Related issues**: Fixes https://github.com/buildbuddy-io/buildbuddy-internal/issues/2234

### CODE REVIEW

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (sluongng) [inline enterprise/server/remote_execution/dirtools/dirtools.go:347]: The current approach is to treat each symlink like a normal file by appending it to `filesToUpload`

In `func uploadFiles()` where we consume this `filesToUpload` slice, we check whether the file entry is a directory or not `uploadableFile.dir == nil` before adding it to our `interfaces.FileCache` where the file would be hardlinked into the cache dir.

Because a symlink is just a file with a different set of permission and content being the Target it should be compatible with FileCache's hardlink approach.

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (bduffany) [reply enterprise/server/remote_execution/dirtools/dirtools.go:412]: If I'm understanding the documentation of `output_symlinks` correctly, it seems like all we need to do is for each `output_path`, if it's a symlink, then just add the symlink metadata (name, target) to `output_symlinks`. I would expect the logic here to just be something like

```
if dirHelper.IsOutputPath(fqfn) {
  actionResult.OutputSymlinks = append(actionResult.OutputSymlinks, &repb.SymlinkNode{Name: trimPathPrefix(fqfn, rootDir), Target: target})
}
```

maybe @iain-macdonald can double-check my understanding since I think he implemented output_paths originally

Comment (sluongng) [reply enterprise/server/remote_execution/dirtools/dirtools.go:412]: https://app.buildbuddy.io/invocation/8e025ac8-fb37-4bcd-97c0-dc4001b64c32?actionDigest=aeb0a9661b70e46c23d9f82317754063a5699656227ab28ee04fb3ceb7e36445#action is an example ActionResult executed locally and cached remotely.

Comment (bduffany) [reply enterprise/server/remote_execution/dirtools/dirtools.go:412]: sorry could you elaborate? I'm not sure what I should be looking for in that invocation link, or how this change is related to local execution

Comment (iain-macdonald) [reply enterprise/server/remote_execution/dirtools/dirtools.go:412]: I don't think my change for output_paths affected the symlink stuff at all, but I can take a closer look if you two need another pair of eyes. Though, I do remember that the change I made broke a project depending on some `rules_kotlin` targets, so I'd recommend that you test the example projects [here](https://github.com/bazelbuild/rules_kotlin/tree/master/examples) against your remote execution changes.

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (bduffany) [inline enterprise/server/remote_execution/dirtools/dirtools.go:367]: The documentation of `output_symlinks` says:

```
// The output paths of the action that are symbolic links to other paths
```

If I'm understanding it correctly, "output paths" is referring to the `output_paths` field. So I don't think `FindParentOutputPath` is the right thing to do here -- instead we only want to set OutputSymlinks for paths that *exactly match* one of the `OutputPaths`.

Example:

```
input:
  - output_paths= [ /foo, /baz ]
action produces:
  /foo (directory)
    /bar (symlink)
  /baz (symlink)
```

I think the expected `OutputSymlinks` should contain just `/baz` since it exactly matches one of the output paths, but I think with this change it will return both `/bar` and `/baz`.

Comment (sluongng) [reply enterprise/server/remote_execution/dirtools/dirtools.go:367]: Yup, that's sensible.

For better context, I was also able to dump out the requested Command from rules_go:
```
{
  "arguments":  [
    "bazel-out/k8-opt-exec-60723E07/bin/go/tools/builders/go_path/go_path-bin",
    "-manifest",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path~manifest",
    "-out",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path",
    "-mode",
    "link"
  ],
  "outputFiles":  [
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/extra.txt",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/pkg/linux_amd64/example.com/repo/cmd/bin.a",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/pkg/linux_amd64/example.com/repo/pkg/lib.a",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/pkg/linux_amd64/example.com/repo/pkg/lib/transitive.a",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/pkg/linux_amd64/example.com/repo/vendor/example.com/repo2.a",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/pkg/linux_amd64/github.com/bazelbuild/rules_go/go/tools/bzltestutil.a",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/pkg/plan9_arm/example.com/repo/cmd/bin.a",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/example.com/repo/cmd/bin/bin.go",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/example.com/repo/pkg/lib/data.txt",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/example.com/repo/pkg/lib/embed_test.go",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/example.com/repo/pkg/lib/embedded_src.txt",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/example.com/repo/pkg/lib/external_test.go",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/example.com/repo/pkg/lib/internal_test.go",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/example.com/repo/pkg/lib/lib.go",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/example.com/repo/pkg/lib/renamed_embedded_src.txt",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/example.com/repo/pkg/lib/template/index.html.tmpl",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/example.com/repo/pkg/lib/testdata/testdata.txt",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/example.com/repo/pkg/lib/transitive/transitive.go",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/example.com/repo/vendor/example.com/repo2/vendored.go",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/github.com/bazelbuild/rules_go/go/tools/bzltestutil/init.go",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/github.com/bazelbuild/rules_go/go/tools/bzltestutil/lcov.go",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/github.com/bazelbuild/rules_go/go/tools/bzltestutil/test2json.go",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/github.com/bazelbuild/rules_go/go/tools/bzltestutil/wrap.go",
    "bazel-out/k8-fastbuild/bin/tests/core/go_path/link_path/src/github.com/bazelbuild/rules_go/go/tools/bzltestutil/xml.go"
  ],
  "platform":  {
    "properties":  [
      {
        "name":  "OSFamily",
        "value":  "Linux"
      },
      {
        "name":  "container-image",
        "value":  "docker://gcr.io/bazel-public/ubuntu1804-bazel-java11@sha256:2d50853a7edbe59a99bc4141d7a03cb1068157b9766077302b46c4ec94eef151"
      }
    ]
  }
}
```

I did not have the time today but tomorrow, I plan to set this up on my Linux laptop and test + iterate from there.

Comment (bduffany) [reply enterprise/server/remote_execution/dirtools/dirtools.go:367]: nice, are you also able to capture the local ActionResult?

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (sluongng) [reply enterprise/server/remote_execution/dirtools/dirtools.go]: Discussed with Brandon on Slack on this and shared some of the initial findings I got by running the build locally against our remote cache.

@iain-macdonald Note on the Kotlin project. Let me try running it against my change.

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (sluongng) [reply enterprise/server/remote_execution/dirtools/dirtools.go]: Was able to use

```
build:buildbuddy --experimental_remote_grpc_log=./grpc.log
```

in combination with `bb print` to gain very useful insight :D

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (sluongng) [reply enterprise/server/remote_execution/dirtools/dirtools.go]: Tested against both `dagger` and `trivial` in rules_kotlin example and they don't have the same problem as rules_go's go_path so both passed with/without this change applied.

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (sluongng) [inline enterprise/server/remote_execution/dirtools/dirtools.go:355]: I was considering adding it here 

```
				if !(dirHelper.ShouldUploadFile(fqfn) || dirHelper.IsOutputPath(fqfn)) {
					continue
				}
```

But I was worried about breaking existing behavior.

It's unclear to me how often people create/include symlink outputs in their actions _without_ declaring them in `Command.output_path`, and if we should start explicitly requiring folks to declare them.

**Comment by sluongng** [comment]
Added some basic test to make this easier to maintain in the long-run 🙏

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (bduffany) [inline enterprise/server/remote_execution/dirtools/dirtools.go:382]: The documentation of `output_symlinks` says:

```
  // New in v2.1: this field will only be populated if the command
  // `output_paths` field was used, and not the pre v2.1 `output_files` or
  // `output_directories` fields.
```

So (1) based on this part: `this field will only be populated if the command `output_paths` field was used`

I think there should be a check around this line like

```
if IsOutputPath(fqfn) {
  actionResult.OutputSymlinks = append(actionResult.OutputSymlinks, ospb)
}
```

However, it looks like `IsOutputPath` is not technically checking the `action.OutputPaths` field. In workspace.go > SetTask, we are setting `outputPaths` to _either_ `output_paths` if set, else `concat(output_files, output_directories)`. This is pretty confusing IMO - we might want to refactor that. But for now we could maybe just pass in the `command` proto so that we have access to the original `output_paths` field.

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (sluongng) [reply enterprise/server/remote_execution/dirtools/dirtools.go:368]: Support for unresolved symlink was added to Bazel in the recent version (since 6.0.0) https://github.com/bazelbuild/bazel/commit/8056eb566371798acdec97aedde0a2c242957ef7
https://github.com/bazelbuild/bazel/issues/10298

When a symlink is unresolved, we cannot tell whether the target is a file or a directory to add it in the respective file/dir ActionResult field.

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (bduffany) [inline enterprise/server/remote_execution/dirtools/dirtools.go:375]:
```suggestion
					// OutputFileSymlinks and OutputDirectorySymlinks are deprecated in REAPI v2.1,
```

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (bduffany) [inline enterprise/server/remote_execution/dirtools/dirtools.go:384]: The documentation of `output_directory_symlinks` says:

```
  // The output directories of the action that are symbolic links to other
  // directories.
```

similarly for `output_file_symlinks`:

```
  // The output files of the action that are symbolic links to other files.
```

I am interpreting this to mean that "the output files of the action" means `Action.command.output_files` and "the output directories of the action" means `Action.command.output_directories`.

If that's correct then under that interpretation, we need to check that this symlink file is exactly referenced in output_files or output_directories before we add them to the ActionResult.output_file_symlinks or ActionResult.output_directory_symlinks (respectively).

(Note that this function begins directory traversal starting from the workspace root, not any of the specific output directories or output file/symlink paths, otherwise we wouldn't need an explicit check.)

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (bduffany) [reply enterprise/server/remote_execution/dirtools/dirtools.go:372]: The current implementation seems fine (as it gives us support for resolved symlinks which we didn't have before), but I am curious how unresolved symlinks can work it we don't set them in the ActionResult.

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (bduffany) [inline enterprise/server/remote_execution/dirtools/dirtools.go:34]: can you revert the noisy diffs to `var` and `0111 -> 0o111` etc (I guess they are caused by gofumpt vs gofmt or something?)

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (bduffany) [inline enterprise/server/remote_execution/dirtools/dirtools.go:372]: uber nit: maybe `symlink` instead of `ospb` which feels a little opaque (what does `pb` stand for?)

File: enterprise/server/remote_execution/dirtools/dirtools_test.go
Comment (bduffany) [inline enterprise/server/remote_execution/dirtools/dirtools_test.go:67]: nit: it might be a little clearer to use `hash.String("a")` from the hash util package

File: enterprise/server/remote_execution/dirtools/dirtools_test.go
Comment (bduffany) [inline enterprise/server/remote_execution/dirtools/dirtools_test.go:81]: I might be misunderstanding but I think an `OutputDirectory` won't include an `OutputFile` as its child, otherwise the child will get uploaded twice. I think a more representative example would be to have "b" as an output directory, then in the fileContents map below, create a few files fileB.txt and fileC.txt

File: enterprise/server/remote_execution/dirtools/dirtools_test.go
Comment (bduffany) [inline enterprise/server/remote_execution/dirtools/dirtools_test.go:99]: how is this TreeDigest computed? IMO it's not easy to verify that this is correct just by looking at it - can we maybe have a util func that computes it?

File: enterprise/server/remote_execution/dirtools/dirtools_test.go
Comment (bduffany) [inline enterprise/server/remote_execution/dirtools/dirtools_test.go:157]: ditto here - I think in practice we wouldn't see `a/fileA.txt` nested under an output directory. (there are also a few more occurrences of this below)

File: enterprise/server/remote_execution/dirtools/dirtools.go
Comment (sluongng) [reply enterprise/server/remote_execution/dirtools/dirtools.go:372]: pb = protobuf... I think I was reusing another variable and never gave it a second thought. 🤣

`symlink` sounds good

File: enterprise/server/remote_execution/dirtools/dirtools_test.go
Comment (sluongng) [reply enterprise/server/remote_execution/dirtools/dirtools_test.go:208]: Added comment.

I think unresolved symlink is only supported as outputs and not as inputs. See https://cs.opensource.google/bazel/bazel/+/master:src/test/java/com/google/devtools/build/lib/remote/RemoteExecutionServiceTest.java;l=216-231?q=RemoteExecution&ss=bazel

Also, the flag is enabled by default since Bazel 6.0.0 so it's not an opt-in feature.

File: enterprise/server/remote_execution/dirtools/dirtools_test.go
Comment (bduffany) [reply enterprise/server/remote_execution/dirtools/dirtools_test.go:98]: Resolving to avoid blocking the PR - might be nice to have a TreeDigest util at some point to make these test assertions a bit clearer.

**Review by bduffany — APPROVED** [review]
Looks great!

**Pull request merged**
